core/grant/service.go

```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
+			uniqueAccountIDsMap[g.AccountID] = true
+		}
+	}
+
+	s.logger.Info(fmt.Sprintf("getting activities for provider %q", provider.URN))
+	activities, err := s.providerService.ListActivities(ctx, *provider, domain.ListActivitiesFilter{
+		AccountIDs:   uniqueAccountIDs,
+		TimestampGte: &startDate,
+	})
+	if err != nil {
+		return fmt.Errorf("listing activities: %w", err)
+	}
+
+	grantsPointer := make([]*domain.Grant, len(grants))
+	for i, g := range grants {
+		g := g
+		grantsPointer[i] = &g
+	}
+	if err := s.providerService.CorrelateGrantActivities(ctx, *provider, grantsPointer, activities); err != nil {
+		return fmt.Errorf("correlating grant activities: %w", err)
+	}
+
+	s.logger.Info("checking grants dormancy...")
+	var dormantGrants []*domain.Grant
+	var dormantGrantsIDs []string
+	var dormantGrantsByOwner = map[string][]*domain.Grant{}
+	for _, g := range grantsPointer {
+		if len(g.Activities) == 0 {
+			g.ExpirationDateReason = fmt.Sprintf("%s: %s", domain.GrantExpirationReasonDormant, criteria.RetainDuration)
+			newExpDate := time.Now().Add(criteria.RetainDuration)
+			g.ExpirationDate = &newExpDate
+
+			dormantGrants = append(dormantGrants, g)
+			dormantGrantsIDs = append(dormantGrantsIDs, g.ID)
+
+			dormantGrantsByOwner[g.Owner] = append(dormantGrantsByOwner[g.Owner], g)
+		}
+	}
+	s.logger.Info(fmt.Sprintf("found %d dormant grants for provider %q", len(dormantGrants), provider.URN), "grant_ids", dormantGrantsIDs)
+
+	if criteria.DryRun {
+		s.logger.Info("dry run mode, skipping updating grants expiration date")
+		return nil
+	}
+
+	if err := s.repo.BulkUpsert(ctx, dormantGrants); err != nil {
+		return fmt.Errorf("updating grants expiration date: %w", err)
+	}
+
+	go func() { // send notifications
+		var notifications []domain.Notification
+	prepare_notifications:
+		for owner, grants := range dormantGrantsByOwner {
+			var grantsMap []map[string]interface{}
+			var grantIDs []string
+
+			for _, g := range grants {
+				grantMap, err := structToMap(g)
+				if err != nil {
+					s.logger.Error("failed to convert grant to map", "error", err)
+					continue prepare_notifications
+				}
+				grantsMap = append(grantsMap, grantMap)
+			}
+
+			notifications = append(notifications, domain.Notification{
+				User: owner,
+				Labels: map[string]string{
+					"owner":     owner,
+					"grant_ids": strings.Join(grantIDs, ", "),
+				},
+				Message: domain.NotificationMessage{
+					Type: domain.NotificationTypeUnusedGrant,
+					Variables: map[string]interface{}{
+						"dormant_grants":       grantsMap,
+						"period":               criteria.Period.String(),
+						"retain_duration":      criteria.RetainDuration.String(),
+						"start_date_formatted": startDate.Format("Jan 02, 2006 15:04:05 UTC"),
+					},
+				},
+			})
+		}
+
+		if errs := s.notifier.Notify(notifications); errs != nil {
+			for _, err1 := range errs {
+				s.logger.Error("failed to send notifications", "error", err1.Error())
+			}
+		}
```

Comment:
Are there test cases for this functionality?

Reply:
test cases added